src/wire/dhcp.rs

```diff
@@ -0,0 +1,629 @@
+pub struct Repr<'a> {
+    /// This field is also known as `op` in the RFC. It indicates the type of DHCP message this
+    /// packet represents.
+    pub message_type: MessageType,
+    /// This field is also known as `xid` in the RFC. It is a random number chosen by the client,
+    /// used by the client and server to associate messages and responses between a client and a
+    /// server.
+    pub transaction_id: u32,
+    /// This field is also known as `chaddr` in the RFC and for networks where the access layer is
+    /// ethernet, it is the client MAC address.
+    pub client_hardware_address: EthernetAddress,
+    /// This field is also known as `ciaddr` in the RFC. It is only filled in if client is in
+    /// BOUND, RENEW or REBINDING state and can respond to ARP requests.
+    pub client_ip: Ipv4Address,
+    /// This field is also known as `yiaddr` in the RFC.
+    pub your_ip: Ipv4Address,
+    /// This field is also known as `siaddr` in the RFC. It may be set by the server in DHCPOFFER
+    /// and DHCPACK messages, and represent the address of the next server to use in bootstrap.
+    pub server_ip: Ipv4Address,
+    /// This field is also known as `giaddr` in the RFC. In order to allow DHCP clients on subnets
+    /// not directly served by DHCP servers to communicate with DHCP servers, DHCP relay agents can
+    /// be installed on these subnets. The DHCP client broadcasts on the local link; the relay
+    /// agent receives the broadcast and transmits it to one or more DHCP servers using unicast.
+    /// The relay agent stores its own IP address in the `relay_agent_ip` field of the DHCP packet.
+    /// The DHCP server uses the `relay_agent_ip` to determine the subnet on which the relay agent
+    /// received the broadcast, and allocates an IP address on that subnet. When the DHCP server
+    /// replies to the client, it sends the reply to the `relay_agent_ip` address, again using
+    /// unicast. The relay agent then retransmits the response on the local network
+    pub relay_agent_ip: Ipv4Address,
+    /// Broadcast flags. It can be set in DHCPDISCOVER, DHCPINFORM and DHCPREQUEST message if the
+    /// client requires the response to be broadcasted.
+    pub broadcast: bool,
+    /// The "requested IP address" option. It can be used by clients in DHCPREQUEST or DHCPDISCOVER
+    /// messages, or by servers in DHCPDECLINE messages.
+    pub requested_ip: Option<Ipv4Address>,
+    /// The "server identifier" option. It is used both to identify a DHCP server
+    /// in a DHCP message and as a destination address from clients to servers.
+    pub server_identifier: Option<Ipv4Address>,
+    pub parameter_request_list: Option<&'a [u8]>,
+}
+
+impl<'a> Repr<'a> {
+    /// Parse a DHCP packet and return a high-level representation.
+    pub fn parse<T>(packet: &Packet<&'a T>) -> Result<Self>
+            where T: AsRef<[u8]> + ?Sized {
+
+        let transaction_id = packet.transaction_id();
```

Comment:
The length of the packet has not been checked before this point right? This could be out of bounds right?

Reply:
Thanks for the review! I'm a bit confused, since my original PR had such a check. It was called a "useless check" [in the previous review](https://github.com/m-labs/smoltcp/pull/75#discussion_r153168767), so I removed it. Should I add it again?

Reply:
When in doubt go with what @whitequark says. I am curious as to why that check is useless though. I would think the following would cause an out of bounds access, but I could be mistaken.

```rust
let bytes = [0x00; 4];
let packet = DhcpPacket::new(&bytes[..])
DhcpRepr::parse(packet)
```